### PR TITLE
Content registry v1

### PR DESCRIPTION
**Pull request by SollyW**
Fixes some issues with, and includes, #576, which the author no longer wants to maintain.

### CODE REVIEW

File: fabric-content-registries-v1/src/main/java/net/fabricmc/fabric/api/content/registry/v1/StrippableBlockRegistry.java [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+
+import net.fabricmc.fabric.api.content.registry.v1.util.ContentRegistry;
+import net.fabricmc.fabric.impl.content.registry.StrippableBlockRegistryImpl;
+
+/**
+ * Registry of Blocks that when stripped turn into a specific Block.
+ * Note: Both KEY and VALUE must have the {@link PillarBlock#AXIS} property!
+ */
+public interface StrippableBlockRegistry {
+	ContentRegistry<Block, Block> INSTANCE = StrippableBlockRegistryImpl.INSTANCE;
```
Comment (i509VCB) [inline fabric-content-registries-v1/src/main/java/net/fabricmc/fabric/api/content/registry/v1/StrippableBlockRegistry.java]: Imo I think these could all become part of a `ContentRegistries` class in API so all these content registries are centrally accessible

File: fabric-content-registries-v0/src/main/java/net/fabricmc/fabric/api/registry/FlammableBlockRegistry.java [hunk 2]
```diff
@@ -21,15 +21,19 @@
...
+ */
+@Deprecated
 public interface FlammableBlockRegistry extends Block2ObjectMap<FlammableBlockRegistry.Entry> {
 	static FlammableBlockRegistry getDefaultInstance() {
 		return getInstance(Blocks.FIRE);
 	}
 
 	static FlammableBlockRegistry getInstance(Block block) {
-		return FlammableBlockRegistryImpl.getInstance(block);
+		return LegacyFlammableBlockRegistryImpl.getInstance(block);
```
Comment (i509VCB) [inline fabric-content-registries-v0/src/main/java/net/fabricmc/fabric/api/registry/FlammableBlockRegistry.java:36]: Is there a way we could inline the old api to call new api/impl?

Comment (SollyW) [reply fabric-content-registries-v0/src/main/java/net/fabricmc/fabric/api/registry/FlammableBlockRegistry.java:36]: See https://github.com/FabricMC/fabric/pull/576/files#r408796138

File: fabric-content-registries-v1/src/main/java/net/fabricmc/fabric/api/content/registry/v1/StrippableBlockRegistry.java (hunk 1 above)
Comment (SollyW) [reply fabric-content-registries-v1/src/main/java/net/fabricmc/fabric/api/content/registry/v1/StrippableBlockRegistry.java]: This has been done. Let me know how it looks

File: fabric-content-registries-v1/src/main/java/net/fabricmc/fabric/api/content/registry/v1/ContentRegistries.java [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+	/**
+	 * Registry of Items to values from 0.0 to 1.0, defining the chance of a given item
+	 * increasing the Composter block's level.
+	 */
+	public static final ItemContentRegistry<Float> COMPOSTABLE_ITEM = CompostableItemRegistryImpl.INSTANCE;
+
+	/**
+	 * Registry of Blocks that can be set on fire by {@link net.minecraft.block.Blocks#FIRE}.
+	 */
+	public static final FlammableBlockRegistry FLAMMABLE_FROM_FIRE = FlammableBlockRegistryImpl.FIRE_INSTANCE;
```
Comment (SollyW) [inline fabric-content-registries-v1/src/main/java/net/fabricmc/fabric/api/content/registry/v1/ContentRegistries.java:45]: Needs a better name

File: fabric-content-registries-v1/src/main/java/net/fabricmc/fabric/api/content/registry/v1/ContentRegistries.java [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+import net.fabricmc.fabric.api.content.registry.v1.util.ContentRegistry;
+import net.fabricmc.fabric.api.content.registry.v1.util.ItemContentRegistry;
+import net.fabricmc.fabric.impl.content.registry.CompostableItemRegistryImpl;
+import net.fabricmc.fabric.impl.content.registry.FlammableBlockRegistryImpl;
+import net.fabricmc.fabric.impl.content.registry.FlattenableBlockRegistryImpl;
+import net.fabricmc.fabric.impl.content.registry.FuelItemRegistryImpl;
+import net.fabricmc.fabric.impl.content.registry.StrippableBlockRegistryImpl;
+import net.fabricmc.fabric.impl.content.registry.TillableBlockRegistryImpl;
+
+public class ContentRegistries {
```
Comment (i509VCB) [inline fabric-content-registries-v1/src/main/java/net/fabricmc/fabric/api/content/registry/v1/ContentRegistries.java]: Add a private constructor please and make class final

Comment (SollyW) [reply fabric-content-registries-v1/src/main/java/net/fabricmc/fabric/api/content/registry/v1/ContentRegistries.java]: Done

File: fabric-content-registries-v1/src/main/java/net/fabricmc/fabric/api/content/registry/v1/ContentRegistries.java (hunk 3 above)
Comment (i509VCB) [reply fabric-content-registries-v1/src/main/java/net/fabricmc/fabric/api/content/registry/v1/ContentRegistries.java:45]: `BURNABLE_BLOCKS`?

File: fabric-content-registries-v1/src/main/java/net/fabricmc/fabric/api/content/registry/v1/FlammableBlockRegistry.java [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+
+	default void add(Block block, int burn, int spread) {
+		this.add(block, new Entry(burn, spread));
+	}
+
+	default void add(Tag<Block> tag, int burn, int spread) {
+		this.add(tag, new Entry(burn, spread));
+	}
+
+	final class Entry {
```
Comment (i509VCB) [inline fabric-content-registries-v1/src/main/java/net/fabricmc/fabric/api/content/registry/v1/FlammableBlockRegistry.java:46]: Entry should be in impl

Comment (Prospector) [reply fabric-content-registries-v1/src/main/java/net/fabricmc/fabric/api/content/registry/v1/FlammableBlockRegistry.java:46]: No, because you can add, get, or remove Entries. Making it an inaccessible class doesn't make sense.

File: fabric-content-registries-v1/src/main/java/net/fabricmc/fabric/api/content/registry/v1/ContentRegistries.java (hunk 3 above)
Comment (Prospector) [reply fabric-content-registries-v1/src/main/java/net/fabricmc/fabric/api/content/registry/v1/ContentRegistries.java:45]: FIRE_FLAMMABLE?

Comment (SollyW) [reply fabric-content-registries-v1/src/main/java/net/fabricmc/fabric/api/content/registry/v1/ContentRegistries.java:45]: Maybe FIRE_FLAMMABLE_BLOCK? The others are all either _BLOCK or _ITEM

Comment (ghost) [reply fabric-content-registries-v1/src/main/java/net/fabricmc/fabric/api/content/registry/v1/ContentRegistries.java:45]: I like i509VCB's suggestion of `BURNABLE_BLOCKS`

Comment (Prospector) [reply fabric-content-registries-v1/src/main/java/net/fabricmc/fabric/api/content/registry/v1/ContentRegistries.java:45]: Flammable > Burnable imo because burnable makes me think more furnace fuel, which is a separate thing.

**Review by Prospector — APPROVED** [review]
I think this could use a test mod, now that those are a thing. I had an example of the tests I used when originally making the API, perhaps you could use those. Other than that, LGTM!

File: fabric-content-registries-v1/src/main/java/net/fabricmc/fabric/api/content/registry/v1/ContentRegistries.java (hunk 3 above)
Comment (SollyW) [reply fabric-content-registries-v1/src/main/java/net/fabricmc/fabric/api/content/registry/v1/ContentRegistries.java:45]: Other registries can exist for other types of fire using this system. That being said, that'd be pretty rare, so something like `FLAMMABLE_BLOCKS` might be good enough

**Comment by SollyW** [comment]
This is still 1.15, so it would have to be added once it gets updated I believe? Anyway, something like [this](https://gist.github.com/Solly-W/2ae7b34fd34a04e9938da18082956faa) should work.

**Comment by i509VCB** [comment]
The test looks fine to me. I'd use the other test mods (see `testmod` source sets for examples how to use testmods) and register it there.

**Comment by SollyW** [comment]
Ah, looks like prospector's fork is just outdated and doesn't have the tests yet

**Comment by i509VCB** [comment]
You might need to pull upstream or rebase to get all the testing stuff up to date.

**Comment by SollyW** [comment]
Smh intellij

File: settings.gradle [hunk 6]
```diff
@@ -19,6 +19,7 @@ include 'fabric-blockrenderlayer-v1'
 include 'fabric-commands-v0'
 include 'fabric-containers-v0'
 include 'fabric-content-registries-v0'
+include 'fabric-content-registries-v1'
```
Comment (i509VCB) [inline settings.gradle:22]: Needs to be renamed to `fabric-content-registry-api-v1` to be in line with new naming standards for modules.

Comment (Prospector) [reply settings.gradle:22]: Actually, per the RFC issue, this would be -registries. Registry-api sounds like an API for adding registries

Comment (i509VCB) [reply settings.gradle:22]: Hmm I see that entry after I looked again at the RFC

File: fabric-content-registries-v1/src/main/java/net/fabricmc/fabric/api/content/registry/v1/ContentRegistries.java [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+import net.fabricmc.fabric.api.content.registry.v1.util.ContentRegistry;
+import net.fabricmc.fabric.api.content.registry.v1.util.ItemContentRegistry;
+import net.fabricmc.fabric.impl.content.registry.CompostableItemRegistryImpl;
+import net.fabricmc.fabric.impl.content.registry.FlammableBlockRegistryImpl;
+import net.fabricmc.fabric.impl.content.registry.FlattenableBlockRegistryImpl;
+import net.fabricmc.fabric.impl.content.registry.FuelItemRegistryImpl;
+import net.fabricmc.fabric.impl.content.registry.StrippableBlockRegistryImpl;
+import net.fabricmc.fabric.impl.content.registry.TillableBlockRegistryImpl;
+
+public final class ContentRegistries {
```
Comment (i509VCB) [inline fabric-content-registries-v1/src/main/java/net/fabricmc/fabric/api/content/registry/v1/ContentRegistries.java:32]: In general I feel this class could use some more documentation

File: fabric-content-registries-v1/src/main/java/net/fabricmc/fabric/api/content/registry/v1/ContentRegistries.java [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+	private ContentRegistries() {
+	}
+
+	/**
+	 * Registry of Items to values from 0.0 to 1.0, defining the chance of a given item
+	 * increasing the Composter block's level.
+	 */
```
Comment (i509VCB) [inline fabric-content-registries-v1/src/main/java/net/fabricmc/fabric/api/content/registry/v1/ContentRegistries.java:39]: I'd reword this a bit to possibly be a bit more clear:

```
A registry of items that may be composed by a {@link ComposterBlock composter}.

<p>When registering an item that is considered compostable, the registry entry specifies a value between {@code 0.0F} and {@code 1.0F} that determines the chance of increasing a composter's {@link Properties#LEVEL_8 level}.

File: fabric-content-registries-v1/src/main/java/net/fabricmc/fabric/impl/content/registry/ContentRegistryImpl.java [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+	}
+
+	@Override
+	public V get(K key) {
+		return getter.apply(key);
+	}
+
+	@Override
+	public void add(K key, V value) {
+		Runnable adder = () -> {
```
Comment (i509VCB) [inline fabric-content-registries-v1/src/main/java/net/fabricmc/fabric/impl/content/registry/ContentRegistryImpl.java:76]: I somewhat dislike the lambda soup and would prefer it to be possible to work on concrete methods.

I seem to see the fact that you have to keep an old copy of the logic without tags which isn't the easiest thing to avoid

File: fabric-content-registries-v1/src/main/java/net/fabricmc/fabric/impl/content/registry/ContentRegistryImpl.java [hunk 10]
```diff
@@ -0,0 +1,143 @@
...
+
+	private boolean tagsPresent = false;
+	private final List<Runnable> processor = new LinkedList<>(); // This is to preserve the order of how things are added and removed. This would be a lot simpler if we did not have to deal with Tags
+	private final Map<K, V> restorer = new HashMap<>(); // Stores the state of the vanilla map without any fabric modifications so everything can be undone on resource reload
+	private final BiConsumer<K, V> putter; // Used to add registry values to whatever stores them
+	private final Consumer<K> remover; // Used to remove registry values from whatever stores them
+	private final Function<K, V> getter; // Used to get registry values from whatever stores them
+
+	protected ContentRegistryImpl(String name, BiConsumer<K, V> putter, Consumer<K> remover, Function<K, V> getter) {
+		reloadIdentifier = new Identifier("fabric:private/" + name);
```
Comment (i509VCB) [inline fabric-content-registries-v1/src/main/java/net/fabricmc/fabric/impl/content/registry/ContentRegistryImpl.java:51]:
```suggestion
		reloadIdentifier = new Identifier("fabric", ":private/" + name);
```

File: fabric-content-registries-v1/src/main/java/net/fabricmc/fabric/impl/content/registry/FlammableBlockRegistryImpl.java [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+import net.minecraft.util.registry.Registry;
+
+import net.fabricmc.fabric.api.content.registry.v1.FlammableBlockRegistry;
+import net.fabricmc.fabric.mixin.content.registry.FireBlockAccessor;
+
+public class FlammableBlockRegistryImpl extends ContentRegistryImpl<Block, FlammableBlockRegistry.Entry> implements FlammableBlockRegistry {
+	public static final FlammableBlockRegistry FIRE_INSTANCE = new FlammableBlockRegistryImpl(Blocks.FIRE, "fire");
+
+	public FlammableBlockRegistryImpl(Block fireBlock, String prefix) {
+		super(prefix + "_flammable_block_registry",
```
Comment (i509VCB) [inline fabric-content-registries-v1/src/main/java/net/fabricmc/fabric/impl/content/registry/FlammableBlockRegistryImpl.java:33]: maybe get the prefix from the block's registry entry?

File: fabric-content-registries-v1/src/main/java/net/fabricmc/fabric/api/content/registry/v1/ContentRegistries.java [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+	public static final ItemContentRegistry<Float> COMPOSTABLE_ITEM = CompostableItemRegistryImpl.INSTANCE;
+
+	/**
+	 * Registry of Blocks that can be set on fire by {@link net.minecraft.block.Blocks#FIRE}.
+	 */
+	public static final FlammableBlockRegistry FLAMMABLE_FROM_FIRE = FlammableBlockRegistryImpl.FIRE_INSTANCE;
```
Comment (i509VCB) [inline fabric-content-registries-v1/src/main/java/net/fabricmc/fabric/api/content/registry/v1/ContentRegistries.java:45]: maybe key this by block instance? And another method to iterate through all blocks with a flammable registry.

This is probably worth the effort since 1.16 adds another fire block

File: fabric-content-registries-v1/src/main/java/net/fabricmc/fabric/impl/content/registry/StrippableBlockRegistryImpl.java [hunk 13]
```diff
@@ -0,0 +1,33 @@
...
+import net.fabricmc.fabric.mixin.content.registry.AxeItemAccessor;
+
+public class StrippableBlockRegistryImpl extends ContentRegistryImpl<Block, Block> {
+	public static final ContentRegistry<Block, Block> INSTANCE = new StrippableBlockRegistryImpl();
+
+	private StrippableBlockRegistryImpl() {
+		super("strippable_block_registry",
+				AxeItemAccessor.getStrippedBlocks()::put, // Putter
+				AxeItemAccessor.getStrippedBlocks()::remove, // Remover
+				AxeItemAccessor.getStrippedBlocks()::get); // Getter
```
Comment (i509VCB) [inline fabric-content-registries-v1/src/main/java/net/fabricmc/fabric/impl/content/registry/StrippableBlockRegistryImpl.java:31]: No validation exists to verify the key and value support the required properties.

**Comment by i509VCB** [comment]
@Solly-W Are you still working on this by chance?

**Comment by Solly-W** [comment]
Yes, I intend to continue working on this soon. Real life has kept me busy the past few weeks.

File: fabric-content-registries-v0/build.gradle [hunk 14]
```diff
@@ -1,5 +1,5 @@
 archivesBaseName = "fabric-content-registries-v0"
-version = getSubprojectVersion(project, "0.1.3")
+version = getSubprojectVersion(project, "0.1.4")
```
Comment (i509VCB) [inline fabric-content-registries-v0/build.gradle:2]: Minor should be getting bumped, not patch

**Comment by i509VCB** [comment]
Effectively superseded by #1125
